### PR TITLE
ADBDEV-4950: Backport a patch for partition locking

### PR DESCRIPTION
**Pull request by bandetto**
Backport a patch for partition locking

A backport (b5c02fa) was merged in upstream, which prevented locks from being
released on partition children or their indexes before the end of transaction,
only for planner. We merged this patch with minor changes (2763ff1) for our
fork.

Then it’s reverted (d40e482) in upstream, but in our fork it’s not. The reason
for revert is that new changes break some test for indexes.

Last month, a different patch (4ac7227) is merged, which supposedly fixes the
parent issue for the planner and ORCA as well. It also does not change the
behavior of locks on indexes, like the first one did. It avoids releasing locks
for AO tables, and hides this ability behind a GUC, because we can exceed the
maximum amount of locks in a single transaction in cases with large partition
hierarchy.

Missing echo of DROP command was added to tests' output of the target patch.

2763ff1, 5894018 and b1373f8 were reverted and re-applied to fix conflicts.

This is a backport of 4ac7227.

### CODE REVIEW

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67146

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67159

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1189782

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67167

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190462

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1190463

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67276

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1195504

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1195505

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/67343

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200086

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1200087

File: src/backend/optimizer/prep/prepunion.c [hunk 1]
```diff
@@ -1558,9 +1560,36 @@ expand_inherited_rtentry(PlannerInfo *root, RangeTblEntry *rte, Index rti)
...
+			 * end of command. Similarly, in the same transaction if we operate
+			 * on these 2 partition hierarchies in any way, we would be holding
+			 * (P1 + P2) locks till the end of the transaction.
+			 */
+			if (RelationIsAppendOptimized(newrelation) &&
+				(parent_is_partitioned || parent_is_interior_partition) &&
+				gp_keep_partition_children_locks)
+				releaseLockMode = NoLock;
+			else
+				releaseLockMode = rel_is_part_child(childOID) ? NoLock: lockmode;
```
Comment (bimboterminator1) [inline src/backend/optimizer/prep/prepunion.c]: Let's preserve the behaviour of DDL patch and keep holding locks as in the patch.

File: src/backend/catalog/dependency.c [hunk 2]
```diff
@@ -1359,6 +1359,21 @@ AcquireDeletionLock(const ObjectAddress *object, int flags)
...
+		 * we are careful to always lock the parent object, we don't need
+		 * to keep the lock on the sub-object. This helps to keep the lock
+		 * table size in check, if you e.g. drop a table with thousands
+		 * of partitions. It would be unpleasent if you could not drop
+		 * such a table because the lock table runs out of space.
+		 *
+		 * To provide at least some protection though, we still actuire
+		 * the lock momentarily.
+		 */
+		if (!gp_keep_partition_children_locks &&
```
Comment (bimboterminator1) [inline src/backend/catalog/dependency.c]: Let's preserve the DDL's patch behaviour. Just provide correct backport of upstream patch for ORCA.

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql [hunk 3]
```diff
@@ -22,6 +23,7 @@ CREATE TABLE ptest(i int) WITH(appendonly=true) PARTITION BY RANGE (i) (START (0
 -- session completes.
 SELECT gp_inject_fault('pg_calculate_table_size_before_pg_aoseg_estimation', 'suspend', dbid)
 FROM gp_segment_configuration WHERE role = 'p' AND content = 0;
+2: SET gp_keep_partition_children_locks TO on;
```
Comment (bimboterminator1) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: Here turning on the GUC value is unnecessary.

**Comment by bimboterminator1** [comment]
Get rid of the conflicts and fix ABI tests

File: src/backend/gpopt/gpdbwrappers.cpp [hunk 4]
```diff
@@ -2736,47 +2739,94 @@ gpdb::IsTypeRange(Oid typid)
 	return false;
 }
 
-RowMarkClause *
-gpdb::GetParseRowmark(Query *query, Index rtindex)
```
Comment (bimboterminator1) [inline src/backend/gpopt/gpdbwrappers.cpp]: We could resolve conflicts without removing the previous code and just change its location inside a file. I guess it is legit to make such changes for **our** patches.  We **do not need** a separate commit 2809cad07593997680149fdffca9fb3f1c89b2b8.

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql [hunk 5]
```diff
@@ -11,6 +11,7 @@ CREATE TABLE ptest(i int) WITH(appendonly=true) PARTITION BY RANGE (i) (START (0
 -- The first session removes partitioned table inside transaction block
 -- When locks on partitions are not set (buggy case), the second session that
 -- works with specific partition will have progress
+1: SET gp_keep_partition_children_locks TO on;
```
Comment (bimboterminator1) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: Do that while applying DDL patch. We do not need this commit

File: src/test/isolation2/expected/concurrent_select_vacuum_ao_partitioned_table.out [hunk 6]
```diff
@@ -233,3 +233,4 @@ VACUUM
  1 | 1 | 2 
 (1 row)
 DROP TABLE foo_part;
+DROP
```
Comment (bimboterminator1) [inline src/test/isolation2/expected/concurrent_select_vacuum_ao_partitioned_table.out:236]: Unfortunately, if the tests do not pass without DROP, we have to add DROP to the **upstream patch**. I don't know why the tests have passed in upstream, but in this PR each separate commit should bring the whole project to the valid state. So, we have no choice but fixing the tests inside the upstream commit.

File: src/backend/optimizer/prep/prepunion.c (hunk 1 above)
Comment (bimboterminator1) [reply src/backend/optimizer/prep/prepunion.c]: @bandetto, I suggest to save the code of DDL patch
https://github.com/arenadata/gpdb/blob/2763ff1446b1e0869c8ad397f3a32204b0c69dda/src/backend/optimizer/prep/prepunion.c#L1562-L1563

Comment (bimboterminator1) [reply src/backend/optimizer/prep/prepunion.c]: @bandetto, let's bring back the DDL's code. Here we'd better remove the rest of the code, and I guess I'll try to introduce this code and the dependency from GUC back in a separate patch later.

File: src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql (hunk 5 above)
Comment (bimboterminator1) [inline src/test/isolation2/sql/ddl_on_root_locks_all_parts.sql]: We don't need to execute this command. We can remove it. And look at the comment about the prepunion code

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68109

**Comment by BenderArenadata** [comment]
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231846

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231843

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231833

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231844

**Comment by BenderArenadata** [comment]
Failed job Regression tests with Postgres on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1231834

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68122

**Comment by BenderArenadata** [comment]
Failed job ORCA code linter on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233712

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233710

**Comment by bimboterminator1** [comment]
@bandetto , Can you fix clang-format tests? One of the patches breaks it.

**Review by bimboterminator1 — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68161

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233899

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233900

**Comment by RekGRpth** [comment]
I think every backport should contain a link/hash of the original commit, as well as a description of what was changed from the original commit and why.

**Comment by BenderArenadata** [comment]
Failed job Regression tests with ORCA on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1233893

**Review by RekGRpth — APPROVED** [review]

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68252

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1237188

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1237189

**Comment by BenderArenadata** [comment]
Allure report https://allure.adsw.io/launch/68448

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1246200

**Comment by BenderArenadata** [comment]
Failed job Resource group isolation tests on ppc64le: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1246201

**Comment by BenderArenadata** [comment]
Failed job Behave tests on x86_64: https://gitlab.adsw.io/arenadata/github_mirroring/gpdb/-/jobs/1246198

**Pull request merged**